### ISSUE TITLE
Owner field is not created if owner belongs to authorised group

### ISSUE THREAD

**Issue by jk171505** [aws-amplify/amplify-category-api#1494]
### How did you install the Amplify CLI?

npm

### If applicable, what version of Node.js are you using?

16.14.0

### Amplify CLI Version

12.0.0

### What operating system are you using?

macos

### Did you make any manual changes to the cloud resources managed by Amplify? Please describe the changes made.

No manual changes made

### Describe the bug

If user belongs to authorised group, the new record is created without an owner field.
I created the schema with the following authorisation rules:
```
type TestMode
  @model
  @auth(
    rules: [
      { allow: private, provider: iam }
      { allow: groups, groups: ["Manager"], groupClaim: "userRoles", provider: oidc }
      { allow: owner, provider: oidc, identityClaim: "sub" }
    ]
  ) {
  id: ID!
  name: String
}
```
If user that creates the record belongs to Manager group, the owner field isn't automatically added to the record.

### Expected behavior

Records are properly attached to the owner regardless of the user group the user belongs to.

### Reproduction steps

1. Create a project with default authorization mode set to OpenID.
2. Issue create mutation in appsync console with authorization provider set as OpenID, providing access token with user that belongs to user group the same as the one set in the schema.

### Log output

<details>

```
# Put your logs below this line


```

</details>

### Before submitting, please confirm:

- [X] I have done my best to include a minimal, self-contained set of instructions for consistently reproducing the issue.
- [X] I have removed any sensitive information from my code snippets and submission.

**Comment by AnilMaktala** [comment]
Hi @jk171505, Thanks for raising this issue. Could you kindly confirm whether you configured the Auth and selected the OpenID connect while creating the API using add API? Additionally, could you please provide the aws-export.js file?

<img width="441" alt="image" src="https://github.com/aws-amplify/amplify-category-api/assets/3206171/8d80efc9-2520-4082-a610-73da417c8e36">

**Comment by jk171505** [comment]
Hi @AnilMaktala, Yes, that's exactly how I set this up.
Where you want me to provide this file?

**Comment by jk171505** [comment]
Hi @AnilMaktala, attached.
[aws-export.js.zip](https://github.com/aws-amplify/amplify-category-api/files/11566703/aws-export.js.zip)

**Comment by alharris-at** [comment]
I've confirmed that expected behavior here SHOULD be that if you're a member of a static group authorizer, and there's an owner field, the owner should be added. This is working with userPools as it should today, and the code for oidc is identical to userPools.

A few things that'll help us get a better bead on what's happening in your specific application. 1) Please run `amplify diagnose --send-report` and provide the diagnostic id in this ticket, and 2) If you're willing, zip up your app client code and email to `amplify-cli@amazon.com` - so we can help inspect the client code, and see if you're not getting the auth mode you'd expect.

The only thing I can think that'd be happening here is that the client request is signed using IAM authorization (instead of OIDC), which is resulting in IAM auth, which will pre-empt the oidc-based auth, and result in no owner being set.

I think there are a few options to debug this @jk171505 - either 1) remove the `{ allow: private, provider: iam }`, and deploy in dev to test out that you're definitely authorizing w/ oidc, (e.g. your app can make a request, that receives a 200 back from the service), 2) in the AWS AppSync console, navigate to the API for this app, and under `Settings/Edit` enable logging, checking 'Include Verbose Content', this will allow you to inspect request logs for headers, which will indicate the auth mode being used. Then you can make a request from your application, and ensure a JWT token (or parsed token w/ claims, etc.) is coming in the app request from AppSync, which should indicate you're authenticated w/ oidc.

In parallel, @AnilMaktala will be attempting to replicate the issue you're experiencing w/ an external OIDC provider.

**Comment by jk171505** [comment]
Hello @alharris-at,

According to the code of the resolver that is generated, what happens is the exact opposite (attachment).
The owner is added only in case user if user is not authorized, and he is in my case authorized on all fields, therefore the part of the code that should set the owner field is never reached...
<img width="651" alt="Screenshot 2023-05-26 at 20 15 41" src="https://github.com/aws-amplify/amplify-category-api/assets/1572947/52b3cfc0-03b9-4e4e-b154-c3b032ea7e8a">

Let me know if sharing my code will be useful or this explains what happens.

**Comment by jk171505** [comment]
@alharris-at @AnilMaktala any update on this?

I did check the logs, and the look fine (identity is properly decoded), but realised that it doesn't matter. The point is that whenever user belongs to "Manager" user group, the owner field isn't attached to this record. The code of the vtl resolver that I passed few days ago clearly confirm this... Looks like a bug to me.

**Comment by alharris-at** [comment]
@jk171505 - I believe this issue has already been addressed. There is a feature flag called `populateownerfieldforstaticgroupauth` which, if set to true, will populate this field on Mutations even if group auth is met.

I've created a sample project matching your input schema, with iam and oidc auth configured, and using admin roles w/ Amplify CLI version `12.1.1`.

Sample auth resolver is below:

```vtl
## [Start] Authorization Steps. **
$util.qr($ctx.stash.put("hasAuth", true))
#set( $inputFields = $util.parseJson($util.toJson($ctx.args.input.keySet())) )
#set( $isAuthorized = false )
#set( $allowedFields = [] )
#if( $util.authType() == "IAM Authorization" )
  #set( $adminRoles = ["sample-admin-role1","admin-2"] )
  #foreach( $adminRole in $adminRoles )
    #if( $ctx.identity.userArn.contains($adminRole) && $ctx.identity.userArn != $ctx.stash.authRole && $ctx.identity.userArn != $ctx.stash.unauthRole )
      #return($util.toJson({}))
    #end
  #end
  #if( $ctx.identity.userArn == $ctx.stash.authRole )
    #set( $isAuthorized = true )
  #end
#end
#if( $util.authType() == "Open ID Connect Authorization" )
  #if( !$isAuthorized )
    #set( $staticGroupRoles = [{"claim":"userRoles","entity":"Manager","allowedFields":["id","name"],"isAuthorizedOnAllFields":true}] )
    #foreach( $groupRole in $staticGroupRoles )
      #set( $groupsInToken = $util.defaultIfNull($ctx.identity.claims.get($groupRole.claim), []) )
      #if( $groupsInToken.contains($groupRole.entity) )
        #if( $groupRole.isAuthorizedOnAllFields )
          #set( $isAuthorized = true )
          #break
        #else
          $util.qr($allowedFields.addAll($groupRole.allowedFields))
        #end
      #end
    #end
  #end
  #set( $ownerEntity0 = $util.defaultIfNull($ctx.args.input.owner, null) )
  #set( $ownerClaim0 = $util.defaultIfNull($ctx.identity.claims.get("sub"), null) )
  #if( !$util.isNull($ownerClaim0) )

    #if( $isAuthorized && $util.isNull($ownerEntity0) && !$ctx.args.input.containsKey("owner") )
      $util.qr($ctx.args.input.put("owner", $ownerClaim0))
    #end
    #if( !$isAuthorized )
      #set( $ownerClaimsList0 = [] )
      #set( $ownerAllowedFields0 = ["id","name"] )
      #set( $isAuthorizedOnAllFields0 = true )
      #if( $ownerClaim0 == $ownerEntity0 || $ownerClaimsList0.contains($ownerEntity0) )
        #if( $isAuthorizedOnAllFields0 )
          #set( $isAuthorized = true )
        #else
          $util.qr($allowedFields.addAll($ownerAllowedFields0))
        #end
      #end
      #if( $util.isNull($ownerEntity0) && !$ctx.args.input.containsKey("owner") )
        $util.qr($ctx.args.input.put("owner", $ownerClaim0))
        #if( $isAuthorizedOnAllFields0 )
          #set( $isAuthorized = true )
        #else
          $util.qr($allowedFields.addAll($ownerAllowedFields0))
        #end
      #end
    #end
  #end
#end
#if( !$isAuthorized && $allowedFields.isEmpty() )
$util.unauthorized()
#end
#if( !$isAuthorized )
  #set( $deniedFields = $util.list.copyAndRemoveAll($inputFields, $allowedFields) )
  #if( $deniedFields.size() > 0 )
    $util.error("Unauthorized on ${deniedFields}", "Unauthorized")
  #end
#end
$util.toJson({"version":"2018-05-29","payload":{}})
## [End] Authorization Steps. **
```

The key difference from your screenshot is that after the !$util.isNull($ownerClaim0) this snippet is included:
```vtl
#if( $isAuthorized && $util.isNull($ownerEntity0) && !$ctx.args.input.containsKey("owner") )
  $util.qr($ctx.args.input.put("owner", $ownerClaim0))
#end
```

which will populate the `owner` field in the resulting record.

I believe enabling that feature flag (and potentially updating your CLI version) should address this issue.

**Comment by jk171505** [comment]
#red@alharris-at did you test it? The create mutation returns null for `owner` field, but the field is created in db. Also the onCreate subscription doesn't fire.

**Comment by jk171505** [comment]
An update to the previous comment:
loos like this partially works: it works when pushed via cli, but when the source environment is merged into some other (i.e. production) branch and deployed via ci/cd (auto-build), [this flag has no effect](https://github.com/aws-amplify/amplify-category-api/issues/1679).
